8359919: Minor java.util.concurrent doc improvements

DougLea opened the pull request:
This collects miscellaneous open issues that can be resolved with documentation updates; each indicated by adding JDK issue numbers

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change requires CSR request [JDK-8360511](https://bugs.openjdk.org/browse/JDK-8360511) to be approved
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issues
 * [JDK-8359919](https://bugs.openjdk.org/browse/JDK-8359919): Minor java.util.concurrent doc improvements (**Bug** - P4)
 * [JDK-8187775](https://bugs.openjdk.org/browse/JDK-8187775): AtomicReferenceFieldUpdater does not support static fields (**Bug** - P4)
 * [JDK-8254060](https://bugs.openjdk.org/browse/JDK-8254060): SubmissionPublisher close hangs if a publication is pending (**Bug** - P4)
 * [JDK-8210149](https://bugs.openjdk.org/browse/JDK-8210149): Example in JavaDoc for java.util.concurrent.Flow violates Reactive Streams spec (**Enhancement** - P4)
 * [JDK-8199501](https://bugs.openjdk.org/browse/JDK-8199501): Improve documentation of CompletableFuture, CompletionStage (**Enhancement** - P4)
 * [JDK-8233050](https://bugs.openjdk.org/browse/JDK-8233050): CompletableFuture `whenComplete` and `thenApply` change exceptional result (**Bug** - P4)
 * [JDK-8210312](https://bugs.openjdk.org/browse/JDK-8210312): JavaDoc example in SubmissionPublisher will potentially crash (**Enhancement** - P4)
 * [JDK-8292365](https://bugs.openjdk.org/browse/JDK-8292365): CompletableFuture and CompletionStage should document Memory Model guarantees (**Enhancement** - P4)
 * [JDK-8356304](https://bugs.openjdk.org/browse/JDK-8356304): Define "enabled" in ScheduledExecutorService (**Bug** - P4)
 * [JDK-8353155](https://bugs.openjdk.org/browse/JDK-8353155): FutureTask#run(): doc implies synchronous, implementation is async (**Enhancement** - P4)
 * [JDK-8186959](https://bugs.openjdk.org/browse/JDK-8186959): Clarify that Executors.newScheduledThreadPool() is fixed-size (**Bug** - P3)
 * [JDK-8190889](https://bugs.openjdk.org/browse/JDK-8190889): TimeUnit.wait should document IllegalMonitorStateException (**Bug** - P4)
 * [JDK-6351533](https://bugs.openjdk.org/browse/JDK-6351533): CyclicBarrier reset() should return the number of awaiters (**Enhancement** - P4)
 * [JDK-6317534](https://bugs.openjdk.org/browse/JDK-6317534): CyclicBarrier should have a cancel() method (**Enhancement** - P3)
 * [JDK-8195628](https://bugs.openjdk.org/browse/JDK-8195628): Documentation for lock(), trylock(), lockInterruptibly​()  of ReentrantReadWriteLock.WriteLock needs to be corrected (**Bug** - P4)
 * [JDK-8333172](https://bugs.openjdk.org/browse/JDK-8333172): Document a recommendation to use VarHandles instead of java.util.concurrent.atomic.*FieldUpdater (**Enhancement** - P4)
 * [JDK-6374942](https://bugs.openjdk.org/browse/JDK-6374942): Improve thread safety of collection .equals() methods (**Bug** - P3)
 * [JDK-7176957](https://bugs.openjdk.org/browse/JDK-7176957): ExecutorService submit method javaDoc enhancement (**Enhancement** - P4)
 * [JDK-8172177](https://bugs.openjdk.org/browse/JDK-8172177): Improve documentation for CompletionException handling (**Bug** - P4)
 * [JDK-6714849](https://bugs.openjdk.org/browse/JDK-6714849): ReentrantReadWriteLock: Abnormal behavior in non-fair mode (**Bug** - P4)
 * [JDK-6625724](https://bugs.openjdk.org/browse/JDK-6625724): Allow ReentrantReadWriteLock to not track per-thread read holds (**Enhancement** - P3)
 * [JDK-6526284](https://bugs.openjdk.org/browse/JDK-6526284): Improve AbstractExecutorService javadoc (**Enhancement** - P4)
 * [JDK-8137156](https://bugs.openjdk.org/browse/JDK-8137156): Javadoc for Future is misleading with respect to cancellation (**Enhancement** - P4)
 * [JDK-6663476](https://bugs.openjdk.org/browse/JDK-6663476): FutureTask.get() may return null if set() is not called from run() (**Bug** - P4)
 * [JDK-8311131](https://bugs.openjdk.org/browse/JDK-8311131): ConcurrentHashMap.forEachKey parallelismThreshold description could be clearer (**Enhancement** - P4)
 * [JDK-8360511](https://bugs.openjdk.org/browse/JDK-8360511): Minor java.util.concurrent doc improvements (**CSR**)


### Reviewers
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/25880/head:pull/25880` \
`$ git checkout pull/25880`

Update a local copy of the PR: \
`$ git checkout pull/25880` \
`$ git pull https://git.openjdk.org/jdk.git pull/25880/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 25880`

View PR using the GUI difftool: \
`$ git pr show -t 25880`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/25880.diff">https://git.openjdk.org/jdk/pull/25880.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/25880#issuecomment-2985235931)
</details>

DougLea commented:
/issue JDK-8187775

DougLea commented:
/issue JDK-8233050

DougLea commented:
/issue JDK-8292365

DougLea commented:
/issue JDK-8356304

pavelrappo commented on src/java.base/share/classes/java/util/concurrent/ForkJoinPool.java:
```diff
@@ -140,7 +140,7 @@
  * that take too long. The scheduled functions or actions may create
  * and invoke other {@linkplain ForkJoinTask ForkJoinTasks}. Delayed
- * actions become <em>enabled</em> and behave as ordinary submitted
+ * actions become <em>enabled</em> for execution and behave as ordinary submitted
```
Does it make sense to retain `<em>`?

pavelrappo commented on src/java.base/share/classes/java/util/concurrent/Flow.java:
```diff
@@ -60,7 +60,8 @@
  * control required in most implementations (for example {@link
- * SubmissionPublisher}).
+ * SubmissionPublisher}), and omits some error processing needed to
+ * fully conform to ReactiveStream rules.
```
I had to re-read the comments on the relevant [issue] from 2018; thanks for fixing it.

Nit: ReactiveStream. Earlier in that doc comment, it is referred to as "reactive-streams". We should choose one of those.

We can say reactive-streams or Reactive Streams. The latter seems to be the term of choice at https://github.com/reactive-streams/reactive-streams-jvm/.

@viktorklang-ora, thoughts?

[issue]: https://bugs.openjdk.org/browse/JDK-8210149

viktorklang-ora replied:
@pavelrappo Indeed, "Reactive Streams" would be better. Perhaps something like this:

```suggestion
 * fully conform to the Reactive Streams specification.
```

viktorklang-ora commented on src/java.base/share/classes/java/util/concurrent/SubmissionPublisher.java:
```diff
@@ -162,8 +162,8 @@
  *   public void onNext(S item) {
- *     subscription.request(1);
  *     submit(function.apply(item));
+ *     subscription.request(1);
```
👍

viktorklang-ora commented on src/java.base/share/classes/java/util/concurrent/atomic/AtomicReferenceFieldUpdater.java:
```diff
@@ -46,7 +46,7 @@
 /**
  * A reflection-based utility that enables atomic updates to
- * designated {@code volatile} reference fields of designated
+ * designated non-static {@code volatile} reference fields of designated
```
Wouldn't this change also apply to the other AtomicXFieldUpdaters?

DougLea replied:
Yes. Will do. Also, as mentioned by @liach we could trap violations by throwing a better exception. But considering that people should be using VarHandles instead anyway these days, and old uses might depend on current behavior, it doesn't seem worthwhile to add? (line 338)
+            if (Modifier.isStatic(modifiers))
+                throw new IllegalArgumentException("Must not be static");

DougLea commented on src/java.base/share/classes/java/util/concurrent/CompletionStage.java:
```diff
@@ -97,18 +97,21 @@
- * given preference and propagated to the dependent stage.
+ * {@link CompletionException} holding the exception as its
+ * cause. This distinguishes exceptions in an action itself from those
+ * it depends on. If you want them handled in the same way, you might
+ * choose to catch {@link RuntimeException}.  If a stage is dependent
```
Done

DougLea commented on src/java.base/share/classes/java/util/concurrent/CompletionStage.java:
```diff
@@ -137,14 +140,20 @@
+ * submission of a computation producing a {@code CompletionStage}
+ * <i>happen-before</I> that computation begins. And actions taken by
+ * a {@code CompletionStage} <i>happen-before</i> actions of any
+ * dependent stage subsequent to its completion.
```
Changed to:
 And actions taken by
 * {@code CompletionStage x} <i>happen-before</i> actions of any
 * dependent stage subsequent to {@code x}'s completion.

DougLea commented:
/issue JDK-8353155

DougLea commented:
/issue JDK-8195628

DougLea commented:
/issue JDK-8333172

AlanBateman commented on src/java.base/share/classes/java/util/concurrent/atomic/AtomicReferenceFieldUpdater.java:
```diff
@@ -68,6 +69,22 @@
+ *     return LEFT.compareAndSet(this, expect, update);
+ *   }
+ *   // ... and so on
+ * }}</pre>
```
I wonder if this should be convert to a `{@snippet ..}` and have it handle the checked exceptions thrown by findVarHandle. I'm just thinking of someone seeing VarHandle usage for first time, then reporting a bug that the example doesn't compile.

DougLea commented on src/java.base/share/classes/java/util/concurrent/SubmissionPublisher.java:
```diff
@@ -602,9 +602,11 @@ public int offer(T item, long timeout, TimeUnit unit,
-     * subscribers have yet completed.
+     * subscribers, and disallows subsequent attempts to publish. To
+     * ensure uniform ordering among subscribers, this method may
+     * await completion of in-progress offer.  Upon return, this
```
"offers" (since no lock ordering guarantees)

DougLea commented on src/java.base/share/classes/java/util/concurrent/locks/ReentrantReadWriteLock.java:
```diff
@@ -791,10 +791,10 @@ public void lockInterruptibly() throws InterruptedException {
         /**
          * Acquires the read lock only if the write lock is not held by
-         * another thread at the time of invocation.
+         * any thread at the time of invocation.
```
Thanks; reverted. One too many find-replaces/

DougLea commented on src/java.base/share/classes/java/util/concurrent/TimeUnit.java:
```diff
@@ -398,6 +398,8 @@ else if (s == MICRO_SCALE)
      * @param timeout the maximum time to wait. If less than
      * or equal to zero, do not wait at all.
+     * @throws  IllegalMonitorStateException  if the current thread is not
+     *               the owner of this object's monitor.
```
thanks. Done.

DougLea commented:
/issue JDK-6374942

DougLea commented:
/issue JDK-7176957

pavelrappo commented on src/java.base/share/classes/java/util/concurrent/CompletionStage.java:
```diff
@@ -137,14 +140,20 @@
  *
+ * <p>Memory consistency effects: Actions in a thread prior to the
+ * submission of a computation producing a {@code CompletionStage}
+ * <i>happen-before</i> that computation begins. And actions taken by
```
Considering that it's the only occurrence of _happen(s)-before_ in this file, it makes sense to link it to its definition similarly to how it is done in other places.

DougLea replied:
Thanks, done.

DougLea commented:
/issue JDK-8172177

DougLea commented:
/issue JDK-6714849

DougLea commented:
/issue JDK-6625724

DougLea commented:
/issue JDK-6526284

liach commented:
Do we need a CSR to check some of these specification updates?

AlanBateman commented:
> Do we need a CSR to check some of these specification updates?

There is at least one new testable assertion in TimeUnit.timedWait that will need a CSR. I'll look out for others.

AlanBateman commented:
I did a pass over the latest changes (9d9239c5) and all looks good. 

There are several classes where the advice for developers could move to an apiNote in the class (or method) description but it would be disruptive given that it's inline with the existing text.

The only change that I think we will need a CSR is the update to TimeUnit as exception is a new testable assertion.

DougLea commented:
Note that  test j/u/c/tck/ testTimeUnit already includes testTimedWait_IllegalMonitorException(). because it was already implied by the spec. The change just makes this explicit. Does this require CSR? (If so, I'd be happy to do it).

AlanBateman commented:
> Note that test j/u/c/tck/ testTimeUnit already includes testTimedWait_IllegalMonitorException(). because it was already implied by the spec. The change just makes this explicit. Does this require CSR? (If so, I'd be happy to do it).

I don't think it's strictly needed here, Joe Darcy is going to give an opinion today.

jddarcy commented:
> > Note that test j/u/c/tck/ testTimeUnit already includes testTimedWait_IllegalMonitorException(). because it was already implied by the spec. The change just makes this explicit. Does this require CSR? (If so, I'd be happy to do it).
> 
> I don't think it's strictly needed here, Joe Darcy is going to give an opinion today.

Okay -- I've taken a quick look over the proposed changes. While most of the changes are clearly informative (and thus don't need a CSR), I think some of them would fall over the line of what needs a CSR (some changes in TimeUnit.java, RunnableFuture.java, Flow.java). Further work might show those other changes and already implied by other parts of the spec, but I think it is less effort overall just to run a quick CSR for this change. Thanks.

AlanBateman reviewed (approved):
Latest version (with the "bulk" update to CHM) looks good.

DougLea commented:
I added CSR. Rather than walking through each of the changes in terms of compatibility, I added comments to the component JBS issues linked in this PR when applicable.
